### ISSUE TITLE
Replaces when upgrading the version of a default provider

### ISSUE THREAD

**Issue by ellismg** [pulumi/pulumi#2753]
We've made changes recently that cause the version number of the underlying resource provider to be [included as part of register resource requests](https://github.com/pulumi/pulumi-terraform/pull/369).

In addition, when creating resources using the default provider, [we now embed a version number in the provider ID, if present](https://github.com/pulumi/pulumi/blob/master/pkg/resource/deploy/providers/provider.go#L61-L86)

Finally, we have a rule in the step generator that when generating steps for an incoming registration, [if the provider has changed, we need to re-create the resource](https://github.com/pulumi/pulumi/blob/master/pkg/resource/deploy/step_generator.go#L271-L272).

Taken together, these changes lead to the following behavior:

1. Deploy a resource using 0.18.3 of `@pulumi/aws`
2. Observe the provider reference in the state file : 

```
            {
                "urn": "urn:pulumi:dev::bug::pulumi:providers:aws::default",
                "custom": true,
                "id": "c98b6b38-e523-4cc8-a0ff-689bdc966db7",
                "type": "pulumi:providers:aws",
                "inputs": {
                    "region": "us-west-2",
                    "version": "0.18.3"
                }
            },
```

3. Update to 0.18.4 of `@pulumi/aws` (the first version to start passing along a version)
4. Run an update, and see that the resource is going to be replaced!

```
Previewing update (dev):

     Type                 Name                            Plan        Info
     pulumi:pulumi:Stack  bug-dev              
 +-  └─ aws:s3:Bucket     my-bucket                       replace     [diff: ~bucket,provider]
```

Exploring the details, we can see that the provider ID has changed (The `bucket` property being changed is an artifact of autonamming):

```
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::bug::pulumi:pulumi:Stack::bug-dev]
    ++aws:s3/bucket:Bucket: (create-replacement)
        [id=my-bucket-17dfe77]
        [urn=urn:pulumi:dev::bug::aws:s3/bucket:Bucket::my-bucket]
        [provider: urn:pulumi:dev::bug::pulumi:providers:aws::default::c98b6b38-e523-4cc8-a0ff-689bdc966db7 => urn:pulumi:dev::bug::pulumi:providers:aws::default_0_18_4::output<string>]
      ~ bucket      : "my-bucket-17dfe77" => "my-bucket-fe67f9c"
    +-aws:s3/bucket:Bucket: (replace)
        [id=my-bucket-17dfe77]
        [urn=urn:pulumi:dev::bug::aws:s3/bucket:Bucket::my-bucket]
        [provider: urn:pulumi:dev::bug::pulumi:providers:aws::default::c98b6b38-e523-4cc8-a0ff-689bdc966db7 => urn:pulumi:dev::bug::pulumi:providers:aws::default_0_18_4::output<string>]
      ~ bucket      : "my-bucket-17dfe77" => "my-bucket-fe67f9c"
    --aws:s3/bucket:Bucket: (delete-replaced)
        [id=my-bucket-17dfe77]
        [urn=urn:pulumi:dev::bug::aws:s3/bucket:Bucket::my-bucket]
```

I do not believe our strategy of forcing replacements like this works with embedding version numbers in our provider IDs.

**Comment by ellismg** [comment]
@swgillespie I think this issue more or less prevents anyone from being able to upgrade from one version of a package to another, if either of them are sending versions.  Do you recall how this is supposed to work?

**Comment by swgillespie** [comment]
Yikes - the issue here is with the name of the provider resource.

This is tricky. On one hand, we *have* to change the name of default providers from just `default`, because it's perfectly legal for there to be multiple instantiated providers for any one particular provider. On the other hand, we can't change the name, because it causes the provider instance to get re-created and all resources replaced. Bumping versions would also change the name of the provider, because the version of the provider is in the URN.

We should talk about this tomorrow - this is thorny.

**Comment by lukehoban** [comment]
I wonder - can we use something similar to the `aliases` feature in https://github.com/pulumi/pulumi/pull/2404 to work around this?

**Comment by swgillespie** [comment]
> I wonder - can we use something similar to the aliases feature in #2404 to work around this?

Maybe? The issue here implies that it's problematic to put the version into the URN, but I don't know how else to name these provider resources. Auto-incrementing integers? Aliasing might help but it doesn't solve the root problem. Providers are identified as a tuple of package name and version and this issue shows that it's not possible to use those two things to derive a URN without breaking the world.

I'm wondering here if we just need to be smarter in the engine whenever we see a change in provider, with the goal of allowing a diff in the `Provider` field if the only change was that a provider's version was updated. A combination of ad-hoc aliasing and a little more smarts in the provider-provider would work around the naming problem.

**Comment by swgillespie** [comment]
How's this for a rule:

"If the engine sees that a resource's Provider field has changed, let O be the old provider and N be the new provider. If both O and N are both default providers, *and there is no diff in their config*, the resource should be treated as if its provider did not change."

I think this covers all of the bases here. In an upgrade scenario, the engine will be able to gracefully recognize the change of provider without trying to replace everything because it knows that there is no reason to do so.

**Comment by ellismg** [comment]
How does this end up interacting with first class providers @swgillespie?  Do we ever end up burning in a version for a first class provider?  If not, what makes them "different" from default providers in our model?  If we do burn in a version, how do we decide what to burn in and would we end up with the same problem when using a first class provider and upgrading from one version to another?

**Comment by swgillespie** [comment]
>  Do we ever end up burning in a version for a first class provider? If not, what makes them "different" from default providers in our model?

Default providers are injected by the source evaluator lazily whenever a resource registration appears that needs one. Providers that are injected in this manner are never first class, since there was no explicit resource registration for them in the source program. Because of the lack of a resource registration, these default providers have no name, and so previously we chose the name "default" for them. Now, the (buggy) behavior assigns them "default" suffixed with their version.

First class providers do not have this problem, because they are explicitly named in the source program, and therefore the engine does not need to make up a name for them.

**Comment by ellismg** [comment]
> First class providers do not have this problem, because they are explicitly named in the source program, and therefore the engine does not need to make up a name for them.

Of course, thanks.

**Comment by juchom** [comment]
Just to let you know that I have this issue with Azure also after updating the npm packages and Pulumi

**Comment by lukehoban** [comment]
We have released `0.17.13` of the Pulumi CLI which includes a fix that will prevent these replaces from being proposed.

**Comment by ellismg** [comment]
Fixed with #2767
